STDEV-6600(deleted)

apliter opened the pull request:
Добавлен код единицы измерения

Commit:
STDEV-6600
Добавлен код единицы измерения

xcam commented on src/main/java/ru/evotor/framework/inventory/InventoryApi.kt:
```diff
@@ -92,7 +92,8 @@ object InventoryApi {
                                         taxNumber = Utils.safeValueOf(TaxNumber::class.java, cursor.getString(cursor.getColumnIndex(ProductTable.ROW_TAX_NUMBER)), TaxNumber.NO_VAT),
-                                        classificationCode = cursor.optString(cursor.getColumnIndex(ProductTable.ROW_CLASSIFICATION_CODE))
+                                        classificationCode = cursor.optString(cursor.getColumnIndex(ProductTable.ROW_CLASSIFICATION_CODE)),
+                                        measureCode = cursor.getInt(cursor.getColumnIndex(ProductTable.ROW_MEASURE_CODE))
```
думаю должен быть optInt?

apliter replied:
measureCode ведь non-nullable 
Зачем opt?

apliter replied:
+

xcam commented on src/main/java/ru/evotor/framework/inventory/ProductMapper.kt:
```diff
@@ -37,7 +37,8 @@ internal object ProductMapper {
                         taxNumber = Utils.safeValueOf(TaxNumber::class.java, cursor.getString(cursor.getColumnIndex(ProductTable.ROW_TAX_NUMBER)), TaxNumber.NO_VAT),
-                        classificationCode = cursor.optString(cursor.getColumnIndex(ProductTable.ROW_CLASSIFICATION_CODE))
+                        classificationCode = cursor.optString(cursor.getColumnIndex(ProductTable.ROW_CLASSIFICATION_CODE)),
+                        measureCode = cursor.getInt(cursor.getColumnIndex(ProductTable.ROW_MEASURE_CODE))
```
думаю должен быть optInt?

apliter replied:
measureCode ведь non-nullable 
Зачем opt?

apliter replied:
+

xcam commented on src/main/java/ru/evotor/framework/kkt/FiscalTags.kt:
```diff
@@ -182,4 +182,9 @@ object FiscalTags {
+    /**
+     * мера количества предмета расчета
+     */
+    const val MEASURE_OF_QUANTITY = 2108
```
везде measureCode, а тут measureOfQuantity. Выберем что-то одно?

apliter replied:
Это же про фискальный тег. Название этого тега - мера количества.
А инвентори нам уже отдает поле measureCode
Так что считаю, что менять не требуется.

xcam replied:
то что нам отдает inventory - это проблема только нас и inventory. Тогда стоит поменять название нашей переменной, чтобы не создавать лишних сущностей

apliter replied:
Поменял на MEASURE_CODE

apliter replied:
+

xcam commented on src/main/java/ru/evotor/framework/inventory/ProductItem.kt:
```diff
@@ -34,6 +34,7 @@ sealed class ProductItem(
             val description: String?,
             val measureName: String,
             val measurePrecision: Int,
+            val measureCode: Int,
```
а не должен ли он быть Int? (с возможностью указать null)

apliter replied:
Нет. Подразуменавется, что код должен быть. 
Если его не отдаст инвентори, я запишу 255

apliter replied:
сделал nullable...

xcam commented on src/main/java/ru/evotor/framework/receipt/Position.java:
```diff
@@ -70,6 +70,11 @@
+     * Код единицы измерения
+     */
+    @FiscalRequisite(tag = FiscalTags.MEASURE_OF_QUANTITY)
+    private int measureCode;
```
а не должен ли он быть Int (с возможностью указать null)

apliter replied:
Нет. Подразуменавется, что код должен быть.
Если его не отдаст инвентори, я запишу 255

apliter replied:
+

xcam replied:
> Нет. Подразуменавется, что код должен быть.

Почему это? У тебя куча старых чеков где его нет и куча интеграций, которые про него ничего не знают

apliter replied:
Поменял на Integer и анотировал как nullable

apliter replied:
+

xcam commented on src/main/java/ru/evotor/framework/receipt/Position.java:
```diff
@@ -786,6 +805,17 @@ private void readAdditionalFields(Parcel in) {
                 readPreferentialMedicine(in);
                 break;
             }
+            case 7: {
```
весь этот switch надо бы переписать на if

apliter replied:
А зачем?

apliter replied:
+

xcam commented on src/main/java/ru/evotor/framework/receipt/Position.java:
```diff
@@ -1160,6 +1192,11 @@ public Builder setMeasurePrecision(int measurePrecision) {
             return this;
         }
 
+        public Builder setMeasureCode(int measureCode) {
```
не пора ли сделать единый сеттер для наименования точности и кода?

apliter replied:
Опять же, зачем? Если по факту все эти три поля используются далеко не всегда вместе.

xcam replied:
ну measurename и measureCode по сути теперь всегда идут вместе

apliter replied:
+

xcam commented on src/main/java/ru/evotor/framework/receipt/ReceiptApi.kt:
```diff
@@ -398,6 +398,7 @@ object ReceiptApi {
                         cursor.getString(cursor.getColumnIndex(PositionTable.COLUMN_NAME)),
                         cursor.getString(cursor.getColumnIndex(PositionTable.COLUMN_MEASURE_NAME)),
                         cursor.getInt(cursor.getColumnIndex(PositionTable.COLUMN_MEASURE_PRECISION)),
+                        cursor.getInt(cursor.getColumnIndex(PositionTable.COLUMN_MEASURE_CODE)),
```
optInt?

apliter replied:
measureCode ведь non-nullable 
Зачем opt?

apliter replied:
+

Commit:
STDEV-6600
Resolve discussions

xcam commented on src/main/java/ru/evotor/framework/KotlinUtils.kt:
```diff
@@ -75,4 +75,21 @@ fun android.database.Cursor.optString(columnIndex: Int): String? {
     return getString(columnIndex)
+}
+
+@Deprecated("Это внутренний метод. Он не должен быть публичным")
```
странно добавлять метод и сразу его Deprecated

xcam commented on src/main/java/ru/evotor/framework/receipt/Position.java:
```diff
@@ -885,6 +884,7 @@ public static Builder newInstance(
                             name,
                             measureName,
                             measurePrecision,
+                            0,
```
кажется ты говорил, что запишешь 255

apliter replied:
В силу изменений выше, поменял тип и также аннотировал как nullable

apliter replied:
по итогу в билдере будет null
+

xcam commented on src/main/java/ru/evotor/framework/receipt/ReceiptApi.kt:
```diff
@@ -398,6 +396,7 @@ object ReceiptApi {
                         cursor.getString(cursor.getColumnIndex(PositionTable.COLUMN_NAME)),
                         cursor.getString(cursor.getColumnIndex(PositionTable.COLUMN_MEASURE_NAME)),
                         cursor.getInt(cursor.getColumnIndex(PositionTable.COLUMN_MEASURE_PRECISION)),
+                        cursor.optInt(cursor.getColumnIndex(PositionTable.COLUMN_MEASURE_CODE)) ?: 255,
```
почему тут 255? 
Во-первых это magic number
Во-вторых ты подменяешь данные: в базе данных у тебя null (так как там не было никаких measurecode)

xcam commented on build.gradle:
```diff
@@ -116,7 +116,7 @@ publishing {
             groupId 'com.github.evotor'
             artifactId "integration-library"
-            version "MRK-166"
+            version "STDEV-6600"
```
мы не должны поднять def version = 15 ?

apliter replied:
+

Commit:
STDEV-6600
Resolve discussions

Commit:
STDEV-6600
Added @Nullable annotation in getter

xcam commented on build.gradle:
```diff
@@ -6,7 +6,7 @@ apply plugin: 'maven-publish'
 
 android {
-    def version = 15
+    def version = 16
```
в посе есть доработки на поддержку новой версии?

apliter replied:
Пока нет, но будут.

apliter replied:
+

xcam commented on src/main/java/ru/evotor/framework/KotlinUtils.kt:
```diff
@@ -75,4 +75,19 @@ fun android.database.Cursor.optString(columnIndex: Int): String? {
     return getString(columnIndex)
+}
+
+fun android.database.Cursor.optInt(columnName: String): Int? {
```
кажется ты не разобрался почему тут был deprecated и просто его удалил. Deprecated был потому, что эти методы имели область видимости public, хотя не являются частью нашего апи. Их нужно было сделать internal или уже, но мы не могли этого сделать, так как потенциально кто-то их уже использует. Отсюда и появился deprecated. Сейчас ты добавляешь новые методы и я хотел бы чтобы мы сразу выставили правильную область видимости

apliter replied:
сделал internal
+

xcam commented on src/main/java/ru/evotor/framework/core/action/datamapper/PositionMapper.java:
```diff
@@ -89,6 +91,7 @@ public static Position from(@Nullable Bundle bundle) {
         String name = bundle.getString(KEY_NAME);
         String measureName = bundle.getString(KEY_MEASURE_NAME);
         int measurePrecision = bundle.getInt(KEY_MEASURE_PRECISION, 0);
+        int measureCode = bundle.getInt(KEY_MEASURE_CODE, 0);
```
у позиции measureCode теперь может быть null. Почему здесь он non null?

apliter replied:
+

xcam commented on src/main/java/ru/evotor/framework/receipt/Position.java:
```diff
@@ -886,6 +886,7 @@ public static Builder newInstance(
                             measureName,
                             measurePrecision,
                             null,
+                            null,
```
может мы должны метод newInstance расширить measureCode?

apliter replied:
+

Commit:
STDEV-6600
Поля позиции, отвечающие за единицу измерения были объединены в одну сущность - Measure.kt

Commit:
STDEV-6600
убрал лишние записи measure как bundle

xcam commented on src/main/java/ru/evotor/framework/core/action/datamapper/PositionMapper.java:
```diff
@@ -142,14 +146,19 @@ public static Position from(@Nullable Bundle bundle) {
+        Measure measure = new Measure(
+                measureName,
+                measurePrecision,
+                measureCode
```
а тут не будет проверки на measureCode == -1?

apliter replied:
+

xcam commented on src/main/java/ru/evotor/framework/receipt/Measure.kt:
```diff
@@ -0,0 +1,87 @@
+) : Parcelable, IBundlable {
+
+    constructor(parcel: Parcel) : this(
+            parcel.readString(),
```
версионирования нет

apliter replied:
+

xcam commented on src/main/java/ru/evotor/framework/receipt/Measure.kt:
```diff
@@ -0,0 +1,87 @@
+
+    companion object {
+
+        fun readFromCursor(cursor: Cursor): Measure? {
```
думаю этого не должно быть в data class

apliter replied:
+

xcam commented on src/main/java/ru/evotor/framework/receipt/Measure.kt:
```diff
@@ -0,0 +1,87 @@
+                precision = 0,
+                code = 0
+        )
+        val CREATOR = object : Parcelable.Creator<Measure> {
```
@JVMStatic не нужен?

apliter replied:
@JvmField
+

xcam commented on src/main/java/ru/evotor/framework/receipt/Position.java:
```diff
@@ -707,8 +683,8 @@ protected Position(Parcel in) {
         this.name = in.readString();
-        this.measureName = in.readString();
-        this.measurePrecision = in.readInt();
+        this.measure.setName(in.readString());
```
NPE?

apliter replied:
добавил инициализацию поля 
Measure.Companion.getDefault();
+

xcam commented on src/main/java/ru/evotor/framework/receipt/Position.java:
```diff
@@ -838,17 +792,25 @@ public Position createFromParcel(Parcel source) {
+            Measure measure = new Measure(
+                    product.getMeasureName(),
+                    product.getMeasurePrecision(),
+                    product.getMeasureCode()
```
странно что в product мы оставили 3 поля, а в позиции 1

apliter replied:
+

Commit:
STDEV-6600
resolve discussions

xcam commented on src/main/java/ru/evotor/framework/receipt/Measure.kt:
```diff
@@ -0,0 +1,84 @@
+    companion object {
+
+        private const val VERSION = 1
+        const val MEASURE_NO_CODE = -1
```
Это кажется не должно быть публичным

xcam commented on src/main/java/ru/evotor/framework/receipt/Measure.kt:
```diff
@@ -0,0 +1,84 @@
+                    measure = Measure(
+                            name = parcel.readString(),
+                            precision = parcel.readInt(),
+                            code = parcel.readValue(Int::class.java.classLoader) as? Int
```
если чтение происходит таким образом (с допуском null), то зачем нам MEASURE_NO_CODE ?

xcam commented on src/main/java/ru/evotor/framework/receipt/Measure.kt:
```diff
@@ -0,0 +1,84 @@
+) : Parcelable, IBundlable {
+
+    override fun writeToParcel(parcel: Parcel, flags: Int) {
+        parcel.writeString(name)
```
а версия и все остальное?

xcam commented on src/main/java/ru/evotor/framework/receipt/Position.java:
```diff
@@ -65,11 +76,7 @@
-     * Точность единицы измерения.
-     */
-    private int measurePrecision;
+    private Measure measure = Measure.Companion.getDefault();
```
что за дефолтные значения?

xcam commented on src/main/java/ru/evotor/framework/receipt/Position.java:
```diff
@@ -1150,16 +1098,6 @@ public Builder setExtraKeys(Set<ExtraKey> extraKeys) {
-            position.measurePrecision = measurePrecision;
-            return this;
-        }
-
```
вместо этого должно появится setMeasure?
